contracts/truefi2/TrueFiPool2.sol

```diff
@@ -0,0 +1,414 @@
+     * @param amount amount of pool tokens to redeem for underlying currencyTokens
+     */
+    function liquidExit(uint256 amount) external {
+        require(block.number != latestJoinBlock[tx.origin], "TrueFiPool: Cannot join and exit in same block");
+        require(amount <= balanceOf(msg.sender), "TrueFiPool: Insufficient funds");
+
+        uint256 amountToWithdraw = poolValue().mul(amount).div(totalSupply());
+        amountToWithdraw = amountToWithdraw.mul(liquidExitPenalty(amountToWithdraw)).div(10000);
+        require(amountToWithdraw <= liquidValue(), "TrueFiPool: Not enough liquidity in pool");
+
+        // burn tokens sent
+        _burn(msg.sender, amount);
+
+        ensureEnoughTokensAreAvailable(amountToWithdraw);
+
+        require(currencyToken.transfer(msg.sender, amountToWithdraw));
+
+        emit Exited(msg.sender, amountToWithdraw);
+    }
+
+    /**
+     * @dev Penalty (in % * 100) applied if liquid exit is performed with this amount
+     * returns 10000 if no penalty
+     */
+    function liquidExitPenalty(uint256 amount) public view returns (uint256) {
+        uint256 lv = liquidValue();
+        uint256 pv = poolValue();
+        if (amount == pv) {
+            return 10000;
+        }
+        uint256 liquidRatioBefore = lv.mul(10000).div(pv);
+        uint256 liquidRatioAfter = lv.sub(amount).mul(10000).div(pv.sub(amount));
+        return uint256(10000).sub(averageExitPenalty(liquidRatioAfter, liquidRatioBefore));
+    }
+
+    /**
+     * @dev Calculates integral of 5/(x+50)dx times 10000
+     */
+    function integrateAtPoint(uint256 x) public pure returns (uint256) {
+        return uint256(ABDKMath64x64.ln(ABDKMath64x64.fromUInt(x.add(50)))).mul(50000).div(2**64);
+    }
+
+    /**
+     * @dev Calculates average penalty on interval [from; to]
+     * @return average exit penalty
+     */
+    function averageExitPenalty(uint256 from, uint256 to) public pure returns (uint256) {
+        require(from <= to, "TrueFiPool: To precedes from");
+        if (from == 10000) {
+            // When all liquid, don't penalize
+            return 0;
+        }
+        if (from == to) {
+            return uint256(50000).div(from.add(50));
+        }
+        return integrateAtPoint(to).sub(integrateAtPoint(from)).div(to.sub(from));
+    }
+
+    /**
+     * @dev Deposit idle funds into strategy
+     * @param amount Amount of funds to deposit into curve
+     */
+    function flush(uint256 amount) external {
+        require(amount <= currencyBalance(), "TrueFiPool: Insufficient currency balance");
+
+        strategy.deposit(amount);
+
+        emit Flushed(amount);
+    }
+
+    /**
+     * @dev Remove liquidity from strategy
+     * @param minTokenAmount minimum amount of tokens to withdraw
+     */
+    function pull(uint256 minTokenAmount) external onlyOwner {
+        // unstake in gauge
+        ensureEnoughTokensAreAvailable(minTokenAmount);
```

Comment:
this function also attempts to withdraw tokens from the strategy, then the next line tries to withdraw again.
I think this is redundant, and can be removed.